### ISSUE TITLE
Fix S6934 FP/FN: attributes inheriting from `RouteAttribute`

### ISSUE THREAD

**Issue by CaringDev** [SonarSource/sonar-dotnet#8985]
### Description

Using attributes inheriting from `RouteAttribute` are not considered for S6934

### Repro steps

```csharp
public sealed class BarRouteAttribute() : RouteAttribute("/barrr/[controller]/[action]");

[BarRoute]
public class SomeController : ControllerBase // FP: S6934
{
    [HttpGet("foo")]
    public string Foo() => "Hi";
}
```

### Expected behavior

No S6934

### Actual behavior

False positive S6934 for controllers decorated with an attribute inheriting from `RouteAttribute`

### Known workarounds

None

### Related information

* SonarAnalyzer.CSharp 9.23.0.88079

**Comment by mary-georgiou-sonarsource** [comment]
Hello @CaringDev, 

Thank you for reporting this. I confirm it's a false positive.

**Comment by mary-georgiou-sonarsource** [comment]
@CaringDev Seeing this example, I'd like to ask what's the use case here so that we can understand better and improve our analyzer.
Why did you create this base class for the attribute?
Thanks!

**Comment by CaringDev** [comment]
We use two attributes for easy, consistent setup of controllers:

```csharp
public sealed class ControllerRouteAttribute() : RouteAttribute("api/v{version:apiVersion}/[controller]");
public sealed class ActionRouteAttribute() : RouteAttribute("api/v{version:apiVersion}/[controller]/[action]");
```

Then, with "find usages", we can immediately see which controllers use "standard" behavior and which ones action method named routing.
We could also use e.g.
```csharp
public static class Routes {
    public const string Controller = "api/v{version:apiVersion}/[controller]";
    public const string Action = "api/v{version:apiVersion}/[controller]/[action]";
}

[Route(Routes.Controller)]
public class FooController : ControllerBase { /* ... */ }
```

The FP should be fixed in any case though... I think there are as many legit use-cases as there are users out there :-)

**Comment by costin-zaharia-sonarsource** [comment]
If I understand correctly, we should actually look for all the attributes that implement `IRouteTemplateProvider`. ASP.NET Core uses the attributes that implement IRouteTemplateProvider to build the initial set of routes.

**Comment by mary-georgiou-sonarsource** [comment]
@CaringDev thanks a lot for the detailed explanation.
One last question: Did you consider the usage of `[Areas](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/areas?view=aspnetcore-8.0)` for this case?

**Comment by CaringDev** [comment]
> Did you consider the usage of [`Areas`](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/areas?view=aspnetcore-8.0) for this case?

No, we have several ASP.Net Core WebAPI (non-MVC) projects... would you think `Area`s would be applicable there as well?
Given our current approach:

```csharp
namespace Container;

[ControllerRoute]
public class FooController : ControllerBase {
    [HttpGet]
    public string Getindex() => "hi";
}

[ActionRoute]
public class BarController : ControllerBase {
   [HttpPost]
   public void ServeDrinks() { }
```

this would involve changing `ControllerRoute` and `ActionRoute` to `Area("Controller")` and `Area("Action")` and adding corresponding conventions, no? I doubt there's a benefit here... or are you thinking of something else?

**Comment by mary-georgiou-sonarsource** [comment]
Thanks a lot again for the detailed reply, @CaringDev!
I did not have something particular in mind; I was just curious if you evaluated the `Areas` feature and, if yes, why you did not go with it.

I guess for this:

> Then, with "find usages", we can immediately see which controllers use "standard" behavior and which ones action method named routing.

Areas might not work (but not sure).
